edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesEnvironment.cs

```diff
@@ -0,0 +1,163 @@
+    using Microsoft.Azure.Devices.Edge.Agent.Docker;
+    using Microsoft.Azure.Devices.Edge.Agent.Docker.Models;
+    using Microsoft.Azure.Devices.Edge.Storage;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Logging;
+
+    /// <summary>
+    /// This implementation gets the module runtime information from KubernetesRuntimeInfoProvider and
+    /// the configuration information from the deploymentConfig.
+    /// TODO: This could be made generic (not docker specific) and moved to Core.
+    /// </summary>
+    public class KubernetesEnvironment : IEnvironment
+    {
+        readonly IRuntimeInfoProvider moduleStatusProvider;
+        readonly IEntityStore<string, ModuleState> moduleStateStore;
+        readonly string operatingSystemType;
+        readonly string architecture;
+        readonly string version;
+        readonly DeploymentConfig deploymentConfig;
+
+        public KubernetesEnvironment(
+            IRuntimeInfoProvider moduleStatusProvider,
+            DeploymentConfig deploymentConfig,
+            IEntityStore<string, ModuleState> moduleStateStore,
+            string operatingSystemType,
+            string architecture,
+            string version)
+        {
+            this.moduleStatusProvider = moduleStatusProvider;
+            this.deploymentConfig = deploymentConfig;
+            this.moduleStateStore = moduleStateStore;
+            this.operatingSystemType = operatingSystemType;
+            this.architecture = architecture;
+            this.version = version;
+        }
+
+        public async Task<ModuleSet> GetModulesAsync(CancellationToken token)
+        {
+            IEnumerable<ModuleRuntimeInfo> moduleStatuses = await this.moduleStatusProvider.GetModules(token);
+            var modules = new List<IModule>();
+            ModuleSet moduleSet = this.deploymentConfig.GetModuleSet();
+
+            foreach (ModuleRuntimeInfo moduleRuntimeInfo in moduleStatuses)
+            {
+                if (moduleRuntimeInfo.Type != "docker" || !(moduleRuntimeInfo is ModuleRuntimeInfo<DockerReportedConfig> dockerRuntimeInfo))
+                {
+                    Events.InvalidModuleType(moduleRuntimeInfo);
+                    continue;
+                }
+
+                if (!moduleSet.Modules.TryGetValue(dockerRuntimeInfo.Name, out IModule configModule) || !(configModule is DockerModule dockerModule))
+                {
+                    dockerModule = new DockerModule(dockerRuntimeInfo.Name, string.Empty, ModuleStatus.Unknown, Core.RestartPolicy.Unknown, new DockerConfig(Constants.UnknownImage, new CreateContainerParameters()), ImagePullPolicy.OnCreate, new ConfigurationInfo(), null);
+                }
+
+                Option<ModuleState> moduleStateOption = await this.moduleStateStore.Get(moduleRuntimeInfo.Name);
+                ModuleState moduleState = moduleStateOption.GetOrElse(new ModuleState(0, moduleRuntimeInfo.ExitTime.GetOrElse(DateTime.MinValue)));
+
```

Comment:
In Kubernetes mode, I'm 99.44% certain the module store is not written to - this is also Health Restart Planner stuff.

Reply:
Today we still the module store to report the Last Exit time and restart count. For now, we don't report these info in K8s.